lll: skip imports

ibez92 opened the pull request:
Skip long line check on imports.


Fixes #3251

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/golangci/golangci-lint?pullRequest=3288) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/golangci/golangci-lint?pullRequest=3288) before we can accept your contribution.<br/><hr/>**Ivan Bezrodnov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/golangci/golangci-lint?pullRequest=3288) it.</sub>

bombsimon reviewed (commented):
Thanks for the PR!

I added some suggestions about simplifying the code and also a question about correctness. Maybe worth adding a unit test for this?

Also I'm curious if we should set `lineNumber := 0` and do `lineNumber++` on line 96. That way you don't need to do that in every new branch you create and just call `continue` since we should always increment the number on each line we scan.

ibez92 commented:
> Thanks for the PR!
> 
> I added some suggestions about simplifying the code and also a question about correctness. Maybe worth adding a unit test for this?
> 
> Also I'm curious if we should set `lineNumber := 1` and do `lineNumber++` on line 96. That way you don't need to do that in every new branch you create and just call `continue` since we should always increment the number on each line we scan.

Thank you for your reply!

I will try to fix all comments.

One question: could you please guide me where should I put unit test for lll? I don't see any *_test file in the golinters directory. Should I put it there?

ldez commented:
@ibez92 you have to sign the CLA

ibez92 commented:
> @ibez92 you have to sign the CLA
> 
> [#3288 (comment)](https://github.com/golangci/golangci-lint/pull/3288#issuecomment-1272601061)

Yeah, missed my email in commit, sry.

Now it's signed

ibez92 commented on pkg/golinters/lll.go:
```diff
@@ -86,9 +90,33 @@ func getLLLIssuesForFile(filename string, maxLineLen int, tabSpaces string) ([]r
+				multiImportEnabled = true
+				lineNumber++
+				continue
+			}
```
You are absolutely right. I removed regexps and now it's looks better.

Thank you!

ibez92 commented on pkg/golinters/lll.go:
```diff
@@ -86,9 +90,33 @@ func getLLLIssuesForFile(filename string, maxLineLen int, tabSpaces string) ([]r
+				continue
+			}
+
+			importsEnded = true
```
Yeah, there was an error. Thank you!

I tried to fix it. But I still have a question about where to put unit test for lll =(

bombsimon commented:
> > Thanks for the PR!
> 
> > 
> 
> > I added some suggestions about simplifying the code and also a question about correctness. Maybe worth adding a unit test for this?
> 
> > 
> 
> > Also I'm curious if we should set `lineNumber := 1` and do `lineNumber++` on line 96. That way you don't need to do that in every new branch you create and just call `continue` since we should always increment the number on each line we scan.
> 
> 
> 
> Thank you for your reply!
> 
> 
> 
> I will try to fix all comments.
> 
> 
> 
> One question: could you please guide me where should I put unit test for lll? I don't see any *_test file in the golinters directory. Should I put it there?

Have a look at [testdata](https://github.com/golangci/golangci-lint/blob/master/test/testdata/lll.go). You can probably just set max lengt in the config to something very short and import any internal package.

ibez92 commented:
> > > Thanks for the PR!
> > 
> > 
> > > 
> > 
> > 
> > > I added some suggestions about simplifying the code and also a question about correctness. Maybe worth adding a unit test for this?
> > 
> > 
> > > 
> > 
> > 
> > > Also I'm curious if we should set `lineNumber := 1` and do `lineNumber++` on line 96. That way you don't need to do that in every new branch you create and just call `continue` since we should always increment the number on each line we scan.
> > 
> > 
> > Thank you for your reply!
> > I will try to fix all comments.
> > One question: could you please guide me where should I put unit test for lll? I don't see any *_test file in the golinters directory. Should I put it there?
> 
> Have a look at [testdata](https://github.com/golangci/golangci-lint/blob/master/test/testdata/lll.go). You can probably just set max lengt in the config to something very short and import any internal package.

Thank you!

I added 2 tests for single and multi import

ldez commented on pkg/golinters/lll.go:
```diff
@@ -86,9 +86,27 @@ func getLLLIssuesForFile(filename string, maxLineLen int, tabSpaces string) ([]r
 
 	lineNumber := 1
 	scanner := bufio.NewScanner(f)
+	multiImportEnabled := false
 	for scanner.Scan() {
 		line := scanner.Text()
 		line = strings.ReplaceAll(line, "\t", tabSpaces)
+		if strings.HasPrefix(line, "import") {
+			if strings.HasSuffix(line, "(") {
+				multiImportEnabled = true
+			}
+
+			lineNumber++
+			continue
+		}
+		if multiImportEnabled {
+			if line == ")" {
+				multiImportEnabled = false
+			}
+
+			lineNumber++
+			continue
+		}
```
```suggestion
	lineNumber := 1
	multiImportEnabled := false

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		line = strings.ReplaceAll(line, "\t", tabSpaces)

		if strings.HasPrefix(line, "import") {
			multiImportEnabled = strings.HasSuffix(line, "(")
		}

		if multiImportEnabled {
			if line == ")" {
				multiImportEnabled = false
			}

			lineNumber++
			continue
		}
```

ibez92 replied:
I removed `if`, but left
```go
lineNumber++
continue
```
for `if strings.HasPrefix(line, "import") {` because we want skip single line import too.

Is it correct?

ldez replied:
yes sorry

ibez92 commented:
> LGTM but can we also add the change to `lineNumber` to avoid the incrementation at three different places? Basically:
> 
> * Set `lineNumber := 0` on line 87
> * Do `lineNumber++` on line 92
> * Remove `lineNumber++` in all branches (line 97, 106 and 121)
> 
> That way we can add any logic and rely on line numbers getting incremented even if we do a `continue` in a branch.

Nice point, thank you!

Did as you suggested

bombsimon reviewed (approved):
LGTM, not merging until you've updated your review status @ldez to avoid miscommunication.

ldez reviewed (approved):
LGTM
